### DISCUSSION TITLE
How to close a connection properly after a @live, @stream or @defer request has closed

### DISCUSSION THREAD

**Discussion by romeovs** [dotansimha/envelop#2119] [General]
How to properly handle the user going away in a `@live`, `@stream` or `@defer` query (or a subscription)?

As far as I can tell there are no hooks to listen to subscriptions or streams closing or stopping.

This makes it hard to properly free up resources (like database connections) when the user goes away.

**Comment by EmrysMyrddin** [comment]
If you want to free up resources, you probably want to execute something after the end of the execution.
This can be done 9with the `ExecuteDoneHook`:

```ts
import { isAsyncIterable, Plugin } from '@envelop/types'

function myCustomPlugin() {
  async function cleanup() {
    // Free up whatever ressources you need here
  }

  return {
    onExeucte: () => ({
      onExecuteDone: result => isAsyncIterable(result) ? { onEnd: cleanup } : cleanup()
    })
  }
}
```
